pkg/koordlet/resmanager/cpu_suppress.go

```diff
@@ -158,98 +159,111 @@ func (r *CPUSuppress) calculateBESuppressCPU(node *corev1.Node, nodeMetric *metr
-func calculateBESuppressCPUSetPolicy(cpusetQuantity *resource.Quantity, oldCPUSetNum int, nodeCPUInfo *metriccache.NodeCPUInfo) []int32 {
-	// set the number of cpuset cpus no less than 2
-	cpus := int32(math.Ceil(float64(cpusetQuantity.MilliValue()) / 1000))
-	if cpus < 2 {
-		cpus = 2
-	}
-
+func calculateBESuppressCPUSetPolicy(cpus int32, processorInfos []util.ProcessorInfo) []int32 {
 	var CPUSets []int32
-
-	numProcessors := int32(len(nodeCPUInfo.ProcessorInfos))
+	numProcessors := int32(len(processorInfos))
 	if numProcessors < cpus {
 		klog.Warningf("failed to calculate a proper suppress policy, available cpus is not enough, "+
 			"please check the related resource metrics: want cpus %v but got %v", cpus, numProcessors)
 		return CPUSets
 	}
 
-	beMaxIncreaseCpuNum := int32(math.Ceil(float64(numProcessors) * beMaxIncreaseCPUPercent))
-	if cpus-int32(oldCPUSetNum) > beMaxIncreaseCpuNum {
-		cpus = int32(oldCPUSetNum) + beMaxIncreaseCpuNum
-	}
-
-	// NOTE: here we prioritize the processors by the affinity of numa node/socket and hyper-threading so that the
-	//       cpuset cpus can be picked in order, rather than scoring all the possible policies.
-	// TBD: may just use a lookup table for a steady state
-
-	// initially sorted by processor id
-	prioritizedCPUs := nodeCPUInfo.ProcessorInfos
-	sort.Slice(prioritizedCPUs, func(i, j int) bool {
-		return prioritizedCPUs[i].CPUID < prioritizedCPUs[j].CPUID
-	})
-
 	// getNodeIndex is a function to calculate an index for every numa node or socket
 	getNodeIndex := func(info util.ProcessorInfo) int32 {
+		// (nodeId, socketId) => nodeIndex
 		return (info.NodeID + numProcessors) * (info.SocketID + 1)
 	}
+	cpuBucketOfNode := map[int32][]util.ProcessorInfo{}
+	for _, p := range processorInfos {
+		nodeIndex := getNodeIndex(p)
+		cpuBucketOfNode[nodeIndex] = append(cpuBucketOfNode[nodeIndex], p)
+	}
 
-	// get the relative indexes of the processor on its node
-	indexOfNode := map[int32]int32{}
-	numProcessorsOfNode := map[int32]int32{}
-	for _, processor := range prioritizedCPUs {
-		numProcessorsOfNode[getNodeIndex(processor)]++
-		if indexOfNode[processor.CoreID] <= 0 {
-			indexOfNode[processor.CoreID] = numProcessorsOfNode[getNodeIndex(processor)]
-		}
+	// change cpuBucket map to array
+	cpuBucket := [][]util.ProcessorInfo{}
+	for _, processorInfos := range cpuBucketOfNode {
+		cpuBucket = append(cpuBucket, processorInfos)
 	}
 
-	// first we prefer the scattered node/socket, secondly the same core
-	// e.g.
-	// for a processorInfo/prioritizedCPUs list:
-	//	CPUID	CoreID	SocketID	NodeID
-	//	0		0		0			0
-	//	1		1		0			0
-	//	2		2		1			1
-	//	3		3		1			1
-	//	4		0		0			0
-	//	5		1		0			0
-	//	6		2		1			1
-	//	7		3		1			1
-	// it will get sorted as below:
-	//	CPUID	CoreID	SocketID	NodeID
-	//	7		3		1			1
-	//	3		3		1			1
-	//	5		1		0			0
-	//	1		1		0			0
-	//	6		2		1			1
-	//	2		2		1			1
-	//	4		0		0			0
-	//	0		0		0			0
```

Comment:
can you keep this kind of figure in comment, so ppl could understand what we're trying to do more easily.